public/services/data_source_service.ts

```diff
@@ -0,0 +1,130 @@
+  DataSourceSelection,
+  Customized,
+}
+
+const getSingleSelectedDataSourceOption = (
+  dataSourceSelection: Map<string, DataSourceOption[]>
+) => {
+  const values = [...dataSourceSelection.values()];
+  // Should use default index if multi data source selected
+  if (values.length === 0 || values.length > 1 || values?.[0]?.length > 1) {
+    return null;
+  }
+  return values[0][0];
+};
+
+export class DataSourceService {
+  private dataSourceId$ = new BehaviorSubject<string | null>(null);
+  private uiSettings: IUiSettingsClient | undefined;
+  private dataSourceManagement: DataSourceManagementPluginSetup | undefined;
+  private dataSourceSelectionSubscription: Subscription | undefined;
+  private dataSourceIdFrom: DataSourceIdFrom | undefined;
+
+  constructor() {}
+
+  initDefaultDataSourceIdIfNeed() {
+    if (!this.isMDSEnabled() || this.dataSourceId$.getValue() !== null) {
+      return;
+    }
+    const defaultDataSourceId = this.uiSettings?.get('defaultDataSource', null);
+    if (!defaultDataSourceId) {
+      return;
+    }
+    this.setDataSourceId(defaultDataSourceId, DataSourceIdFrom.UiSettings);
+  }
+
+  clearDataSourceId() {
+    this.setDataSourceId(null, undefined);
+  }
+
+  getDataSourceQuery() {
+    if (!this.isMDSEnabled()) {
+      return {};
+    }
+    const dataSourceId = this.dataSourceId$.getValue();
+    if (dataSourceId === null) {
+      throw new Error('No data source id');
+    }
+    if (dataSourceId === '') {
+      return {};
+    }
+    return { dataSourceId };
+  }
+
+  isMDSEnabled() {
+    return !!this.dataSourceManagement;
+  }
+
+  setDataSourceId(newDataSourceId: string | null, dataSourceIdFrom: DataSourceIdFrom | undefined) {
+    this.dataSourceIdFrom = dataSourceIdFrom;
+    if (this.dataSourceId$.getValue() === newDataSourceId) {
+      return;
+    }
+    this.dataSourceId$.next(newDataSourceId);
+  }
+
+  getDataSourceId() {
+    return this.dataSourceId$.getValue();
+  }
+
+  subscribeDataSourceId(observer?: PartialObserver<string | null>) {
+    return this.dataSourceId$.subscribe(observer);
+  }
+
+  setup({
+    uiSettings,
+    dataSourceManagement,
+  }: {
+    uiSettings: IUiSettingsClient;
+    dataSourceManagement?: DataSourceManagementPluginSetup;
+  }) {
+    this.uiSettings = uiSettings;
+    this.dataSourceManagement = dataSourceManagement;
+
+    this.dataSourceSelectionSubscription = this.dataSourceManagement?.dataSourceSelection
+      ?.getSelection$()
+      .subscribe((dataSourceSelection) => {
+        const selectedDataSourceOption = getSingleSelectedDataSourceOption(dataSourceSelection);
+        this.setDataSourceId(
+          selectedDataSourceOption?.id ?? null,
+          DataSourceIdFrom.DataSourceSelection
+        );
+      });
+
+    this.uiSettings.get$('defaultDataSource', null).subscribe((newDataSourceId) => {
+      if (this.dataSourceIdFrom === DataSourceIdFrom.UiSettings) {
+        this.setDataSourceId(newDataSourceId, DataSourceIdFrom.UiSettings);
```

Comment:
Q: why do we need to subscribe to the change of `defaultDataSource`? My understanding is `this.setDataSourceId` should be setting the current selected data source, it should not take `defaultDataSource` into consideration.

I guess what we are trying to implement here is when `this.dataSourceId$` is empty, assistant should use `defaultDataSource`, maybe we should incorporate this logic in `getDataSourceId` and `subscribeDataSourceId`?